File: src/ipc/message_queue.py
"""
CPJ IPC Module - Python ZeroMQ Integration
"""
import zmq
import threading
import queue
import time
from enum import Enum
from typing import Optional, Union

class MessageQueueMode(Enum):
    PUBLISHER = 1
    SUBSCRIBER = 2
    REQUEST = 3
    REPLY = 4

class MessageQueue:
    def __init__(self, mode: MessageQueueMode, address: str):
        self.mode = mode
        self.address = address
        self.context = zmq.Context()
        self.message_queue = queue.Queue()
        self.running = False
        
        if mode == MessageQueueMode.PUBLISHER:
            self.socket = self.context.socket(zmq.PUB)
            self.socket.bind(address)
        elif mode == MessageQueueMode.SUBSCRIBER:
            self.socket = self.context.socket(zmq.SUB)
            self.socket.connect(address)
            self.socket.setsockopt_string(zmq.SUBSCRIBE, '')
            self._start_receive_thread()
        elif mode == MessageQueueMode.REQUEST:
            self.socket = self.context.socket(zmq.REQ)
            self.socket.connect(address)
        elif mode == MessageQueueMode.REPLY:
            self.socket = self.context.socket(zmq.REP)
            self.socket.bind(address)
            self._start_receive_thread()
    
    def __del__(self):
        self.close()
    
    def close(self):
        """Clean up resources"""
        if hasattr(self, 'running') and self.running:
            self.running = False
            if hasattr(self, 'receive_thread'):
                self.receive_thread.join()
        if hasattr(self, 'socket'):
            self.socket.close()
        if hasattr(self, 'context'):
            self.context.term()
    
    def publish(self, message: Union[str, bytes]):
        """Publish a message (PUBLISHER mode only)"""
        if self.mode != MessageQueueMode.PUBLISHER:
            raise RuntimeError("Can only publish from PUBLISHER socket")
        
        if isinstance(message, str):
            message = message.encode('utf-8')
        self.socket.send(message)
    
    def send(self, message: Union[str, bytes]):
        """Send a message (REQUEST mode only)"""
        if self.mode != MessageQueueMode.REQUEST:
            raise RuntimeError("Can only send from REQUEST socket")
        
        if isinstance(message, str):
            message = message.encode('utf-8')
        self.socket.send(message)
    
    def receive(self, timeout: Optional[float] = None) -> str:
        """Receive a message"""
        try:
            message = self.message_queue.get(timeout=timeout)
            if isinstance(message, bytes):
                return message.decode('utf-8')
            return message
        except queue.Empty:
            return None
    
    def has_messages(self) -> bool:
        """Check if there are messages available"""
        return not self.message_queue.empty()
    
    def _start_receive_thread(self):
        """Start the background receiving thread"""
        self.running = True
        self.receive_thread = threading.Thread(target=self._receive_loop)
        self.receive_thread.daemon = True
        self.receive_thread.start()
    
    def _receive_loop(self):
        """Background thread for receiving messages"""
        while self.running:
            try:
                if self.socket.poll(timeout=100, flags=zmq.POLLIN):
                    message = self.socket.recv()
                    self.message_queue.put(message)
            except zmq.ZMQError:
                if not self.running:
                    break
                time.sleep(0.001)

# Example usage
if __name__ == '__main__':
    import time
    
    # Create publisher and subscriber
    publisher = MessageQueue(MessageQueueMode.PUBLISHER, "tcp://127.0.0.1:5555")
    subscriber = MessageQueue(MessageQueueMode.SUBSCRIBER, "tcp://127.0.0.1:5555")
    
    # Wait for connection to establish
    time.sleep(0.1)
    
    # Test pub/sub
    publisher.publish("Hello ZeroMQ!")
    time.sleep(0.1)  # Wait for message to be received
    
    message = subscriber.receive()
    print(f"Received: {message}")
    
    # Clean up
    publisher.close()
    subscriber.close()

File: java/src/main/java/cpj/ipc/MessageQueue.java
package cpj.ipc;

import org.zeromq.SocketType;
import org.zeromq.ZMQ;
import org.zeromq.ZContext;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.core.JsonProcessingException;

public class MessageQueue implements AutoCloseable {
    public enum Mode {
        PUBLISHER,
        SUBSCRIBER,
        REQUEST,
        REPLY
    }

    private final Mode mode;
    private final String address;
    private final ZContext context;
    private final ZMQ.Socket socket;
    private final BlockingQueue<String> messageQueue;
    private final AtomicBoolean running;
    private Thread receiveThread;
    private final ObjectMapper jsonMapper = new ObjectMapper();

    public MessageQueue(Mode mode, String address) {
        this.mode = mode;
        this.address = address;
        this.context = new ZContext();
        this.messageQueue = new LinkedBlockingQueue<>();
        this.running = new AtomicBoolean(false);

        switch (mode) {
            case PUBLISHER:
                socket = context.createSocket(SocketType.PUB);
                socket.bind(address);
                break;
            case SUBSCRIBER:
                socket = context.createSocket(SocketType.SUB);
                socket.connect(address);
                socket.subscribe("".getBytes());
                startReceiveThread();
                break;
            case REQUEST:
                socket = context.createSocket(SocketType.REQ);
                socket.connect(address);
                break;
            case REPLY:
                socket = context.createSocket(SocketType.REP);
                socket.bind(address);
                startReceiveThread();
                break;
            default:
                throw new IllegalArgumentException("Invalid mode: " + mode);
        }
    }

    public void publish(String message) {
        if (mode != Mode.PUBLISHER) {
            throw new IllegalStateException("Can only publish from PUBLISHER socket");
        }
        socket.send(message.getBytes(ZMQ.CHARSET));
    }

    public void publishObject(Object obj, String targetLang) throws Exception {
        if (mode != Mode.PUBLISHER) {
            throw new IllegalStateException("Can only publish from PUBLISHER socket");
        }
        String jsonStr = TypeMarshaller.marshal(obj, obj.getClass().getSimpleName(), targetLang);
        socket.send(jsonStr.getBytes(ZMQ.CHARSET));
    }

    public void send(String message) {
        if (mode != Mode.REQUEST) {
            throw new IllegalStateException("Can only send from REQUEST socket");
        }
        socket.send(message.getBytes(ZMQ.CHARSET));
    }

    public String receive() throws InterruptedException {
        return messageQueue.take();
    }

    public String receive(long timeout) throws InterruptedException {
        String message = messageQueue.poll(timeout, java.util.concurrent.TimeUnit.MILLISECONDS);
        return message;
    }

    public <T> T receiveObject(Class<T> type) throws Exception {
        String message = messageQueue.take();
        return TypeMarshaller.unmarshal(message, type);
    }

    public <T> T receiveObject(Class<T> type, long timeout) throws InterruptedException, JsonProcessingException {
        String message = receive(timeout);
        if (message == null)
            return null;
        return jsonMapper.readValue(message, type);
    }

    public boolean hasMessages() {
        return !messageQueue.isEmpty();
    }

    private void startReceiveThread() {
        running.set(true);
        receiveThread = new Thread(() -> {
            while (running.get()) {
                byte[] message = socket.recv(ZMQ.DONTWAIT);
                if (message != null) {
                    messageQueue.offer(new String(message, ZMQ.CHARSET));
                }
                try {
                    Thread.sleep(1);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        });
        receiveThread.setDaemon(true);
        receiveThread.start();
    }

    @Override
    public void close() {
        running.set(false);
        if (receiveThread != null) {
            System.out.println("Closing message queue at address: " + address);
            receiveThread.interrupt();
            try {
                receiveThread.join(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
        socket.close();
        context.close();
    }

    public static void main(String[] args) {
        try {
            // Create publisher and subscriber
            MessageQueue publisher = new MessageQueue(Mode.PUBLISHER, "tcp://127.0.0.1:5556");
            MessageQueue subscriber = new MessageQueue(Mode.SUBSCRIBER, "tcp://127.0.0.1:5556");

            // Wait for connection to establish
            Thread.sleep(100);

            // Test pub/sub
            publisher.publish("Hello from Java ZeroMQ!");
            Thread.sleep(100); // Wait for message to be received

            String message = subscriber.receive(1000);
            System.out.println("Received: " + message);

            // Clean up
            publisher.close();
            subscriber.close();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}